### PR TITLE
Send GOAWAY to server on Client Transport Shutdown

### LINKED ISSUE #460: Send GOAWAY on client shutdown
The HTTP/2 spec encourages sending GOAWAY before dropping the connection. GOAWAY is necessary for "graceful" shutdowns.

When error handling:

> Endpoints SHOULD send a GOAWAY frame when ending a connection, providing that circumstances permit it.

In the GOAWAY frame description section:

> Endpoints SHOULD always send a GOAWAY frame before closing a connection so that the remote peer can know whether a stream has been partially processed or not.

.

> A GOAWAY frame might not immediately precede closing of the connection; a receiver of a GOAWAY that has no more use for the connection SHOULD still send a GOAWAY frame before terminating the connection.

When discussing connection management:

> When either endpoint chooses to close the transport-layer TCP connection, the terminating endpoint SHOULD first send a GOAWAY (Section 6.8) frame so that both endpoints can reliably determine whether previously sent frames have been processed and gracefully complete or terminate any necessary remaining tasks.

### PR DESCRIPTION
**Pull request by aranjans**
Fixes #460 

RELEASE NOTES:

* client: client sends GOAWAY to server on graceful shutdown

### CODE REVIEW

**Comment by ginayeh** [comment]
@aranjans can you check test failures?

File: internal/transport/http2_client.go [hunk 1]
```diff
@@ -983,6 +1027,18 @@ func (t *http2Client) Close(err error) {
 	}
 	t.mu.Unlock()
 	t.controlBuf.finish()
+	t.controlBuf.setError(nil)
+	err1 := t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("GOAWAY from client"), headsUp: true})
```
Comment (arvindbr8) [inline internal/transport/http2_client.go:1031]: After calling `controlBuf.finish()`, a `put` would just be a no-op. I think there are a couple of options here. 

One other way I can think about is, if we can make sure the loopywriter,run() has returned, it should be safe to access the framer directly and write the goAway frame directly. 

Or the other option is to somehow make sure we `put()` the GoAway and also assure that this is the last thing that is processed in the controlbuf.

Comment (arvindbr8) [reply internal/transport/http2_client.go:1031]: setError(nil) is only required since finish() sets the error which causes the put to be a no-op. I would prefer taking another approach.

File: internal/transport/transport_test.go [hunk 2]
```diff
@@ -1188,6 +1188,88 @@ func (s) TestServerConnDecoupledFromApplicationRead(t *testing.T) {
 
 }
 
+// Test that a transport level client shutdown successfully sends a GOAWAY frame
+// to underlying connection.
+func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:1193]: We might have discussed this offline about writing a test in the end2end test package using the clienttester to test this behavior. Might be worth to explore that.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
I've taken a first pass at the PR. Let me know if you have any questions

**Comment by arvindbr8** [comment]
Also the tests are failing due to a panic on a nil pointer reference. Seems like its the `goAwaySent` event that was added.

File: internal/transport/http2_client.go (hunk 1 above)
Comment (aranjans) [reply internal/transport/http2_client.go]: Sure, I have updated the PR

**Comment by shashkin** [comment]
@aranjans vet-proto test should be fixed after #7028 so you need to pull updated master

File: test/goaway_test.go [hunk 3]
```diff
@@ -761,3 +761,50 @@ func (s) TestTwoGoAwayPingFrames(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Failed to dial: %v", err)
+	}
+
+	st := newServerTesterFromConn(t, conn)
+	st.greet()
+	goAwayFrameReceived := testutils.NewChannel()
+	go func() {
+		for {
+			f, err := st.readFrame()
```
Comment (arvindbr8) [inline test/goaway_test.go:785]: Wouldnt this race with the server created above?

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
The changes to transport looks right to me. However I think we are missing something in the test. Could you take another look?

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
The changes to transport looks right to me. However I think we are missing something in the test. Could you take another look?

File: test/goaway_test.go [hunk 4]
```diff
@@ -761,3 +761,50 @@ func (s) TestTwoGoAwayPingFrames(t *testing.T) {
 		t.Fatalf("Error waiting for graceful shutdown of the server: %v", err)
 	}
 }
+
+func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (arvindbr8) [reply test/goaway_test.go:765]: clienttester might be missing a few methods, which you might need for this.

File: test/goaway_test.go [hunk 5]
```diff
@@ -761,3 +761,60 @@ func (s) TestTwoGoAwayPingFrames(t *testing.T) {
...
+				t.Logf("server tester received unexpected frame type %T", f)
+			}
+		}
+	}()
+
+	// Wait for the GOAWAY frame or timeout
+	select {
+	case <-goAwayFrameReceived:
+		t.Logf("GOAWAY frame received by server")
+	case <-time.After(5 * time.Second): // Adjust timeout if needed
```
[the remark this answers is not in the document]
Comment (arvindbr8) [reply test/goaway_test.go:817]: using timer like this might leak resources. You should either create a timer and defer a cancel on it. Or simply use context with timeout.

File: test/goaway_test.go [hunk 6]
```diff
@@ -761,3 +761,60 @@ func (s) TestTwoGoAwayPingFrames(t *testing.T) {
...
+	conn, err := net.DialTimeout("tcp", lis.Addr().String(), defaultTestTimeout)
+	if err != nil {
+		t.Fatalf("Error dialing: %v", err)
+	}
+	defer conn.Close()
+
+	// Create a helper to read frames (similar to your serverTester)
+	st := newServerTesterFromConn(t, conn)
+
+	// Channel to signal if the GOAWAY frame was received
```
Comment (arvindbr8) [inline test/goaway_test.go:794]: Prefer using complete sentences. i.e., punctuate the comment like an actual sentence (like terminating with a period).

Comment (arvindbr8) [reply test/goaway_test.go:794]: See. https://google.github.io/styleguide/go/decisions#doc-comments

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
I think we need to iterate over the test once again. Left a few comments.

File: internal/transport/http2_client.go [hunk 7]
```diff
@@ -454,13 +454,21 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr resolver.Address, opts
 	}
 
 	t.connectionID = atomic.AddUint64(&clientConnectionCounter, 1)
+	t.loopy = newLoopyWriter(clientSide, t.framer, t.controlBuf, t.bdpEst, t.conn, t.logger)
+	t.loopy.lWriterDone = grpcsync.NewEvent()
```
Comment (arvindbr8) [inline internal/transport/http2_client.go:458]: 2 comments, 

1. Why not init `lWriterDone` inside of `newLoopyWriter`
2. I think we can get away without the lWriterDone
  - Seems like this channel is required for blocking the client from calling conn.Close before we return from the loopywriter. We already have writeDone channel which is used for the same purpose.

File: internal/transport/transport_test.go [hunk 8]
```diff
@@ -2664,3 +2664,104 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+		}
+	}()
+
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		return (&net.Dialer{}).DialContext(ctx, "tcp", addr)
+	}
+	copts := ConnectOptions{
+		Dialer:         dialer,
+		ChannelzParent: channelz.RegisterSubChannel(-1, "test subchannel"),
+	}
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2743]: copts is not necessary for this test.

File: internal/transport/transport_test.go [hunk 9]
```diff
@@ -2664,3 +2664,104 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+	ct, err := NewClientTransport(connectCtx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, copts, func(GoAwayReason) {})
+	if err != nil {
+		t.Fatalf("Error while creating client transport: %v", err)
+	}
+	callHdr := &CallHdr{
+		Host:   "localhost",
+		Method: "foo.Small",
+	}
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2751]: callHdr is also not required to be set for this test.

File: internal/transport/transport_test.go [hunk 10]
```diff
@@ -2664,3 +2664,104 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+		t.Fatalf("wrong stream id: %d", s1.id)
+	}
+	ct.Close(errors.New("manually closed by client"))
+	t.Logf("Closed the client connection")
+	e, err := success.Receive(connectCtx)
+	if e != nil || err != nil {
+		t.Fatalf("Error in frame received: %v. Error receiving from channel: %v", e, err)
+	} else {
+		t.Logf("Server received the GOAWAY from client")
+	}
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2766]: we should prolly have two channels? one for success and the other for error?

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Nice work on getting it this far. I think we should be able to wrap it up in a couple of iterations. I've commented an alternative. We can also discuss in detail offline.

File: internal/transport/transport_test.go [hunk 11]
```diff
@@ -2664,3 +2664,91 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+			if goAwayFrame.ErrCode == http2.ErrCodeNo && string(goAwayFrame.DebugData()) == "graceful_shutdown" {
+				t.Logf("Received goAway frame from client")
+				close(successCh)
+			} else {
+				close(errorCh)
+			}
+			return
+		default:
+			// The client should have sent any frame other than GOAWAY
+			t.Logf("The server received a frame other than GOAWAY")
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2728]: Seems like this is also a failure path? We should `close(errorCh)` here as well?

File: internal/transport/http2_client.go [hunk 12]
```diff
@@ -991,7 +1003,10 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
-	t.controlBuf.finish()
+	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("graceful_shutdown"), closeConnErr: errors.New("graceful_shutdown")})
```
Comment (arvindbr8) [inline internal/transport/http2_client.go:1006]: It would be nice if we could send more info in the goAway about the graceful shutdown


```suggestion
	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte(fmt.Sprintf("client shutdown with: %v", err)), closeConnErr: err})
```

File: internal/transport/transport_test.go [hunk 13]
```diff
@@ -2664,3 +2664,91 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+			return
+		}
+	}()
+
+	ct, err := NewClientTransport(connectCtx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{}, func(GoAwayReason) {})
+	if err != nil {
+		t.Fatalf("Error while creating client transport: %v", err)
+	}
+	s1, err1 := ct.NewStream(connectCtx, &CallHdr{})
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2737]: nit:
```suggestion
	s1, err := ct.NewStream(connectCtx, &CallHdr{})
```

File: internal/transport/controlbuf.go [hunk 14]
```diff
@@ -190,10 +190,10 @@ type incomingGoAway struct {
...
 
 type goAway struct {
-	code      http2.ErrCode
-	debugData []byte
-	headsUp   bool
-	closeConn error // if set, loopyWriter will exit, resulting in conn closure
+	code         http2.ErrCode
+	debugData    []byte
+	headsUp      bool
+	closeConnErr error // if set, loopyWriter will exit, resulting in conn closure
```
Comment (arvindbr8) [inline internal/transport/controlbuf.go:196]: The comment here isn't accurate. wdyt?

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

File: internal/transport/http2_client.go [hunk 15]
```diff
@@ -991,7 +1003,10 @@ func (t *http2Client) Close(err error) {
...
+	if t.loopy != nil {
```
[the remark this answers is not in the document]
Comment (arvindbr8) [reply internal/transport/http2_client.go:1007]: Also seems like this read here causes a data race

File: internal/transport/transport_test.go [hunk 16]
```diff
@@ -2664,3 +2664,87 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+		}
+		sfr := http2.NewFramer(sconn, sconn)
+		if err := sfr.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
+			return
+		}
+
+		sfr.ReadFrame()
+		sfr.ReadFrame()
+		sfr.ReadFrame()
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2705]:
```suggestion
		fr, _ := sfr.ReadFrame()
		if _, ok := fr.(*http2.SettingsFrame); !ok {
			t.Errorf("Expected settings frame, got %v", fr)
		}
		fr, _ = sfr.ReadFrame()
		if fr, ok := fr.(*http2.SettingsFrame); !ok && fr.IsAck() {
			t.Errorf("Expected settings ACK frame, got %v", fr)
		}
		fr, _ = sfr.ReadFrame()
		if fr, ok := fr.(*http2.HeadersFrame); !ok && fr.Flags.Has(http2.FlagHeadersEndStream) {
			t.Errorf("Expected Headers frame with END_HEADERS frame, got %v", fr)
		}
```

File: internal/transport/transport_test.go [hunk 17]
```diff
@@ -2664,3 +2664,87 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+// clientTransport.Close(), client sends a goaway to the server with the correct
+// error code and debug data.
+func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// greetDone channel verifies that greet is done with the connection
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2678]: nit: the channel is "used" to verify. So i might rephrase like
```suggestion
	// greetDone is used to notify when server is done greeting the client.
```

and for the following ones.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

File: internal/transport/http2_client.go [hunk 18]
```diff
@@ -468,6 +472,7 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr resolver.Address, opts
 			t.conn.Close()
 		}
 		close(t.writerDone)
+		t.writerMu.Unlock()
```
[the remark this answers is not in the document]
Comment (aranjans) [reply internal/transport/http2_client.go:475]: Wouldn't that cause deadlock? Let's say we acquire lock to initialise it and we release it. Now it acquires lock for [this](https://github.com/grpc/grpc-go/pull/7015/files#diff-5334018444bcb34d15bcb3c8a3050cdf9c8a40fe41305c3a602ea4bfeb0cd178R1013) and since the lock is already acquired, it will never execute [this](https://github.com/grpc/grpc-go/pull/7015/files#diff-5334018444bcb34d15bcb3c8a3050cdf9c8a40fe41305c3a602ea4bfeb0cd178R474)

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Looked at the changes more closely. 

The reason why we had to do this nil check [here](https://github.com/aranjans/grpc-go/blob/71a70f7d05e613a4c18f47d4fa1c0c1b82c50ece/internal/transport/http2_client.go#L1013) is to handle the case when t.Close() is called before we create the goroutine which creates the loopy. And it is because of this check we are having to use a mutex to handle synchronization of the channel. 

Instead we could just ensure that we dont block read on writerDone channel before calling Close. And according to me I think [this](https://github.com/aranjans/grpc-go/blob/71a70f7d05e613a4c18f47d4fa1c0c1b82c50ece/internal/transport/http2_client.go#L415) defer is the only place where Close is called before the loopy is created. 

I think it might help making `writerDone` a `grpcsync.Event`, which should make it easier to close the channel. It might be tricky otherwise with dealing with panics of closing a closed channel.

Also per our discussion from easier, try to see if its easier to do the same test in a more end2end style.

**Comment by arvindbr8** [comment]
Discussed with @aranjans offline. We figured a way to do it without the grpcsync.Event. @aranjans let me know when the PR is ready for another review.

File: internal/transport/http2_client.go [hunk 19]
```diff
@@ -991,7 +1003,14 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
-	t.controlBuf.finish()
+	// The HTTP/2 spec mentions that a GOAWAY frame should be sent before a connection close. If this close() function
+	// ever starts to take in an HTTP/2 error code the peer will be able to get more information about the reason
+	// behind the connection close.
+	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte(fmt.Sprintf("client shutdown with: %v", err)), closeConnErr: err})
+
```
Comment (arvindbr8) [inline internal/transport/http2_client.go:1010]: please nix this \n

File: internal/transport/transport_test.go [hunk 20]
```diff
@@ -2664,3 +2664,95 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+		if _, err := io.ReadFull(sconn, make([]byte, len(clientPreface))); err != nil {
+			t.Errorf("Error while writing settings ack: %v", err)
+			return
+		}
+		sfr := http2.NewFramer(sconn, sconn)
+		if err := sfr.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
+			return
+		}
+
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2702]: please nix this newline. I think the general philosophy should be a newline separator for logical blocks.

File: internal/transport/transport_test.go [hunk 21]
```diff
@@ -2664,3 +2664,95 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+		}
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
+			// Records that the server successfully received a GOAWAY frame.
+			goAwayFrame := fr
+			if goAwayFrame.ErrCode == http2.ErrCodeNo {
+				t.Logf("Received goAway frame from client")
+				close(successCh)
+			} else {
+				close(errorCh)
```
Comment (arvindbr8) [inline internal/transport/transport_test.go:2729]: We should log the unexpected situation maybe by signalling the error back to the read. 

The philosophy here should be that when this test fails later in a regression, it should be clear for the user from the logs to understand what failed.

**Review by arvindbr8 — APPROVED** [review]
LGTM modulo the few comments. I'm ccing @dfawley here for a second look at this patchset.

File: internal/transport/http2_client.go [hunk 22]
```diff
@@ -469,6 +471,7 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr resolver.Address, opts
 		}
 		close(t.writerDone)
 	}()
+
```
Comment (arvindbr8) [inline internal/transport/http2_client.go:474]: nit: i would remove this newline 

```suggestion
```

File: internal/transport/http2_client.go [hunk 23]
```diff
@@ -991,7 +1002,13 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
-	t.controlBuf.finish()
+	// The HTTP/2 spec mentions that a GOAWAY frame should be sent before a connection close. If this close() function
+	// ever starts to take in an HTTP/2 error code the peer will be able to get more information about the reason
+	// behind the connection close.
```
Comment (arvindbr8) [inline internal/transport/http2_client.go:1007]: nit: most of the comment is actually obvious. Calling out an excerpt from the spec might be more useful here. how about? 
```suggestion
	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
```

File: internal/transport/controlbuf.go [hunk 24]
```diff
@@ -190,10 +190,10 @@ type incomingGoAway struct {
...
 
 type goAway struct {
-	code      http2.ErrCode
-	debugData []byte
-	headsUp   bool
-	closeConn error // if set, loopyWriter will exit, resulting in conn closure
+	code         http2.ErrCode
+	debugData    []byte
+	headsUp      bool
+	closeConnErr error // if set, loopyWriter will exit with this error
```
Comment (arvindbr8) [inline internal/transport/controlbuf.go:196]: Still this is only true on the server side. Is this true on the client side as well?

Comment (aranjans) [reply internal/transport/controlbuf.go:196]: Yeah this is valid for client side as well. We are looping out from writer when we receive this error.

File: internal/transport/http2_client.go [hunk 25]
```diff
@@ -991,7 +1002,12 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
-	t.controlBuf.finish()
+	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
+	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte(fmt.Sprintf("client shutdown with: %v", err)), closeConnErr: err})
+	if t.writerDone != nil {
```
Comment (dfawley) [inline internal/transport/http2_client.go:1008]: When is `writerDone == nil`?

File: internal/transport/controlbuf.go (hunk 24 above)
Comment (dfawley) [inline internal/transport/controlbuf.go:196]: Maybe just revert this change since it results in churn and isn't a substantial improvement.

File: internal/transport/transport_test.go [hunk 26]
```diff
@@ -2664,3 +2664,95 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// greetDone is used to notify when server is done greeting the client.
+	greetDone := make(chan struct{})
+	// successCh verifies that GOAWAY is received at server side
+	successCh := make(chan struct{})
+	// errorCh verifies that desired GOAWAY not received by server
+	errorCh := make(chan struct{})
```
Comment (dfawley) [inline internal/transport/transport_test.go:2683]: These can probably be combined..  The pattern is usually:

```go
errCh := make(chan error)
...
// on error:
errCh<-err
close(errCh)

// on success:
close(errCh)
```

File: test/goaway_test.go [hunk 27]
```diff
@@ -761,3 +761,66 @@ func (s) TestTwoGoAwayPingFrames(t *testing.T) {
...
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	val, err := ctCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("timeout waiting for client transport (should be given after http2 creation)")
+	}
+	ct := val.(*clientTester)
+	goAwayReceived := make(chan struct{})
+	errCh := make(chan struct{})
```
Comment (dfawley) [inline test/goaway_test.go:797]: As above, you can just use 1 channel for both purposes.

File: internal/transport/http2_client.go [hunk 28]
```diff
@@ -517,6 +519,15 @@ func (t *http2Client) getPeer() *peer.Peer {
 	}
 }
 
+// OutgoingGoAwayHandler writes GOAWAY to the connection.  Always returns (false, err) as we want the GoAway
+// to be the last frame loopy writes to the transport.
+func (t *http2Client) outgoingGoAwayHandler(g *goAway) (bool, error) {
+	if err := t.framer.fr.WriteGoAway(math.MaxUint32, http2.ErrCodeNo, g.debugData); err != nil {
```
Comment (dfawley) [inline internal/transport/http2_client.go:525]: This should write the max stream ID and not maxint.

**Review by dfawley — COMMENTED** [review]
Looks good overall.  Thanks for fixing this ancient issue!

File: internal/transport/http2_client.go [hunk 29]
```diff
@@ -408,10 +408,9 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr resolver.Address, opts
 	readerErrCh := make(chan error, 1)
 	go t.reader(readerErrCh)
 	defer func() {
-		if err == nil {
-			err = <-readerErrCh
-		}
 		if err != nil {
+			// writerDone should be closed since the loopy goroutine wouldn't have started in the case this function returns an error.
```
Comment (dfawley) [inline internal/transport/http2_client.go:412]: Please wrap all comments at 80 columns (there's more below, I believe).

File: internal/transport/http2_client.go [hunk 30]
```diff
@@ -991,7 +1003,10 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
-	t.controlBuf.finish()
+	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
+	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte(fmt.Sprintf("client shutdown with: %v", err)), closeConn: err})
```
Comment (dfawley) [inline internal/transport/http2_client.go:1008]: Let's omit the client shutdown error from the data we send to the server.  It doesn't seem like a good idea to send that.  It's probably just `ErrConnClosing`, but better safe than sorry.

File: internal/transport/transport_test.go [hunk 31]
```diff
@@ -2664,3 +2664,98 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
+			// Records that the server successfully received a GOAWAY frame.
+			goAwayFrame := fr
+			if goAwayFrame.ErrCode == http2.ErrCodeNo {
+				t.Logf("Received goAway frame from client")
+				close(errorCh)
+			} else {
+				t.Logf("Received unexpected goAway frame from client")
+				errorCh <- errors.New("received unexpected goAway frame from client")
```
Comment (dfawley) [inline internal/transport/transport_test.go:2727]: To be more helpful these need to include the ErrCode received (and also the one expected)

File: internal/transport/transport_test.go [hunk 32]
```diff
@@ -2664,3 +2664,98 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+				close(errorCh)
+			}
+			return
+		default:
+			t.Logf("The server received a frame other than GOAWAY")
+			errorCh <- errors.New("server received a frame other than GOAWAY")
+			close(errorCh)
```
Comment (dfawley) [inline internal/transport/transport_test.go:2734]: Similar as above.

File: internal/transport/transport_test.go [hunk 33]
```diff
@@ -2664,3 +2664,98 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+	t.Logf("Closed the client connection")
+	select {
+	case err = <-errorCh:
+	case <-ctx.Done():
+		t.Errorf("Timed out")
+	}
+	if err != nil {
+		t.Errorf("Error receiving the GOAWAY frame: %v", err)
+	} else {
+		t.Logf("Received a GOAWAY frame from client")
```
Comment (dfawley) [inline internal/transport/transport_test.go:2759]: You log this above, so no need to repeat it here.

File: internal/transport/transport_test.go [hunk 34]
```diff
@@ -2664,3 +2664,98 @@ func TestConnectionError_Unwrap(t *testing.T) {
...
+	_, err = ct.NewStream(ctx, &CallHdr{})
+	if err != nil {
+		t.Fatalf("failed to open stream: %v", err)
+	}
+	// Wait until server receives the headers and settings frame as part of greet.
+	<-greetDone
+	ct.Close(errors.New("manually closed by client"))
+	t.Logf("Closed the client connection")
+	select {
+	case err = <-errorCh:
```
Comment (dfawley) [inline internal/transport/transport_test.go:2752]: Simplify to:

```go
select {
case err := <-errorCh:
	t.Errorf(...)
}
```

please.

File: test/goaway_test.go [hunk 35]
```diff
@@ -761,3 +761,66 @@ func (s) TestTwoGoAwayPingFrames(t *testing.T) {
...
+				close(errCh)
+			}
+		}
+	}()
+	cc.Close()
+	defer ct.conn.Close()
+	select {
+	case <-goAwayReceived:
+	case <-errCh:
+		t.Errorf("Error receiving the goAway: %v", err)
```
Comment (dfawley) [inline test/goaway_test.go:822]: You never send the error on the channel or read it here, so this `err` will be `nil` (from the last `ctCh.Receive`).

File: internal/transport/http2_client.go (hunk 30 above)
Comment (dfawley) [reply internal/transport/http2_client.go]: This still includes the error.  I would prefer if there was no dynamic data in here, just a static string that says the client is shutting down and closing the connection.

File: internal/transport/transport_test.go (hunk 34 above)
Comment (dfawley) [reply internal/transport/transport_test.go:2750]: `:=` please, to limit scope

File: test/goaway_test.go [hunk 36]
```diff
@@ -818,7 +820,7 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 	defer ct.conn.Close()
 	select {
 	case <-goAwayReceived:
-	case <-errCh:
+	case err = <-errCh:
```
Comment (dfawley) [inline test/goaway_test.go:823]: `:=` please, to limit scope

File: internal/transport/http2_client.go (hunk 30 above)
Comment (aranjans) [reply internal/transport/http2_client.go]: I think this is already fixed, but i am not sure why github doesn't shows the change made post the comment.

Comment (dfawley) [reply internal/transport/http2_client.go]: I don't know, but please fix it.

**Review by dfawley — APPROVED** [review]
LGTM modulo the requested change to the last stream ID field in the GOAWAY.

Thanks and congrats on your first PR!

**Review by arvindbr8 — APPROVED** [review]
LGTM. Congrats on the first PR!
